llvm/lib/Target/RISCV/RISCVGISel.td

```diff
@@ -50,6 +50,11 @@ def GIAddrRegImm :
   GIComplexOperandMatcher<s32, "selectAddrRegImm">,
   GIComplexPatternEquiv<AddrRegImm>;
 
+def GIVLOpS32 : GIComplexOperandMatcher<s32, "selectVLOp">,
+                GIComplexPatternEquiv<VLOp>;
+def GIVLOpS64 : GIComplexOperandMatcher<s64, "selectVLOp">,
```

Comment:
I'm not sure how this works. The pattern refers to `VLOp`. How does tablegen know which of these GISel rules to pick?

Reply:
My understanding is that it'll try to match s32, and if it succeeds it will do the RISCVInstructionSelector::selectVLOp in place of `RISCVDAGToDAGISel::selectVLOp`, which is associated with `def VLOp : ComplexPattern<XLenVT, 1, "selectVLOp">;`. If it fails, it will repeat for i64. Only one will be true at a time because the operand cannot be both s32 and s64.

But maybe someone else can weigh in here, because I am not too familiar with how this works.

Reply:
My question is how it works in this tablegen code which can only remember one of the values.

```
  for (const Record *Equiv :                                                     
       RK.getAllDerivedDefinitions("GIComplexPatternEquiv")) {                   
    const Record *SelDAGEquiv = Equiv->getValueAsDef("SelDAGEquivalent");        
    if (!SelDAGEquiv)                                                            
      continue;                                                                  
    ComplexPatternEquivs[SelDAGEquiv] = Equiv;                                   
  }
```

Reply:
I've added tests to show that it is working as expected for rv32 and rv64.

Reply:
What if you delete one of them? Does it still work?

Reply:
It works if I delete one of them. As we discussed offline, maybe the whole type thing isn't actually implemented correctly. I've updated the patch to call this out.

Reply:
Pretty sure it works because there is an explicit cast around the `VLOp` part in this pattern `def VLOpFrag : PatFrag<(ops), (XLenVT (VLOp (XLenVT AVL:$vl)))>`